### PR TITLE
Enable AOT compatibility for all libraries

### PR DESCRIPTION
**Pull request by eerhardt**
Enable AOT compatibility for all libraries and fix AOT warnings.

- Enable configuration binder source generator
- The only library that can't use the ConfigBinder SG is the HeaderParsing library.
  - Blocked by https://github.com/dotnet/runtime/issues/94547

When we get a build of .NET 8 with all the source generator bugs fixed, I will port this to `main`.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4625)

### CODE REVIEW

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Workarounds.cs
Comment (eerhardt) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Workarounds.cs:9]:
```suggestion
// needed to work around https://github.com/dotnet/runtime/issues/94065
global using System.Net.Http;
global using System.Threading.Tasks;
global using Polly.Hedging;
global using Polly.RateLimiting;
global using Polly.Retry;
global using Polly.Timeout;
```

**Comment by RussKie** [comment]
@eerhardt I suggest reducing the threshold in the csproj. The coverage report is showing 100%, but it's likely bugged (we likely need to use the newer PublishCodeCoverageResults task, but it's broken too).

**Comment by eerhardt** [comment]
> @eerhardt I suggest reducing the threshold in the csproj. The coverage report is showing 100%, but it's likely bugged (we likely need to use the newer PublishCodeCoverageResults task, but it's broken too).

I'm assuming it is this change that is confusing/breaking it:

image

I'll change MinCodeCoverage to 99 for this assembly.

File: eng/packages/General.props
Comment (eerhardt) [reply eng/packages/General.props:15]: Fixed.

File: src/Libraries/Microsoft.Extensions.AmbientMetadata.Application/Microsoft.Extensions.AmbientMetadata.Application.csproj
Comment (eerhardt) [reply src/Libraries/Microsoft.Extensions.AmbientMetadata.Application/Microsoft.Extensions.AmbientMetadata.Application.csproj:9]: I wouldn't recommend it for now. This source generator uses the experimental "Interceptors" feature in C# 12. That's why it isn't enabled by default everywhere. Plus we are only using it in 10 of the ~35 projects. So I don't think it makes sense to turn it on for all. Just like we don't turn on the Logging and Metrics source generators globally.

**Comment by eerhardt** [comment]
FYI - @ericstj @eiriktsarpalis @tarekgh - This is changing dotnet/extensions to use the configuration binder source generator in 9.0. No new issues were found. The only outstanding issues that I know about are:

* [Configbinder SG doesn't work with StringValues](https://github.com/dotnet/runtime/issues/94547)
* [Support TimeSpan type in the RangeAttribute in options validation SG](https://github.com/dotnet/runtime/issues/94119)
  * This issue isn't highlighted here because the AOT Analyzer doesn't analyze source generated code. But if you use `Microsoft.Extensions.Http.Resilience` in an AOT app, you'll get 3 warnings from this generated code.

**Comment by eerhardt** [comment]
@joperezr @RussKie @sebastienros - any thoughts on this PR?

**Review by RussKie — APPROVED** [review]
LGTM (with my minimal knowledge of the area)

**Comment by RussKie** [comment]
I'm happy to merge it if you think it's ready.

**Comment by eerhardt** [comment]
> I'm happy to merge it if you think it's ready.

I'm going to run CI on it one more time just to make sure no breaks came in the meantime. Once that passes, I believe this can be merged.

**Pull request merged**
